### PR TITLE
Update swift-wisdom to Swift 3

### PR DESCRIPTION
**Pull request by brightredchilli**
This PR encompasses all changes to make Swift Wisdom compatible with Swift 3. Thanks everyone who helped out! The goal of this PR is to get a workable swift wisdom so that people can start using it in their projects. Therefore, there remains some TODOs and some inconsistencies. This PR includes changes from nyc/standard-library, whose commits have been reviewed. This PR also includes two _unreviewed_ commits:
- [a bit more bookkeeping](https://github.com/IntrepidPursuits/swift-wisdom/commit/db14e47e78f7a370edec95cd7cc5921a75c78ad5)
- [attempt to rename some more classes](https://github.com/IntrepidPursuits/swift-wisdom/commit/1b910f7733e6e9df0566f7761c0f7e3c8c99da58)

This commits can be reviewed individually.

Other than that, all tests are passing, and I have tested this against a swift-3 project. :tada: To try it out, add this to your Podfile:

```
pod 'Intrepid', :git => 'https://github.com/IntrepidPursuits/swift-wisdom.git', :tag => '0.6.0'
```

### CODE REVIEW

File: SwiftWisdom/Core/Async/Qu/Qu.swift [hunk 1]
```diff
@@ -165,38 +164,39 @@ extension Operation {
...
         }
         addOperation(blockOp)
         return blockOp
     }
-}
\ No newline at end of file
+
+    // MARK: Operators
+
+    static func +=(operationQueue: OperationQueue, block: @escaping Block) {
```
Comment (paulrolfe) [inline SwiftWisdom/Core/Async/Qu/Qu.swift]: Our style guide states there should be spaces after the operators and before the first `(` https://github.com/IntrepidPursuits/swift-style-guide#operator-definitions

File: SwiftWisdom/Core/Async/Qu/Qu.swift [hunk 2]
```diff
@@ -165,38 +164,39 @@ extension Operation {
...
+        operationQueue.addOperation(block)
+    }
+    static func +=(operationQueue: OperationQueue, operation: Foundation.Operation) {
```
Comment (paulrolfe) [inline SwiftWisdom/Core/Async/Qu/Qu.swift]: spacing here too

File: SwiftWisdom/Core/DirectoryManager/DirectoryManager.swift [hunk 3]
```diff
@@ -10,96 +10,102 @@ import Foundation
...
 public final class DirectoryManager {
     
-    enum Error : ErrorType {
+    enum DirectoryError : Error {
         // TODO: When upgrading to 2.3 Swift, this type changed from NSURL -> NSURL?. 
         // In swift 2.3, there are more cases in which NSURL creation can fail resulting in an error. 
         // The existing UnableToCreatePath error was modified to capture this case rather than creating a new url-less error case.
-        case UnableToCreatePath(url: NSURL?) 
+        case unableToCreatePath(url: NSURL)
+        case unableToFindPath(url: NSURL)
```
Comment (paulrolfe) [inline SwiftWisdom/Core/DirectoryManager/DirectoryManager.swift]: It seems that this enum case isn't used anywhere.

File: SwiftWisdom/Core/Geometry/Geometry.swift [hunk 4]
```diff
@@ -38,15 +38,15 @@ public extension CGPoint {
...
+    public func ip_distance(to point: CGPoint) -> CGFloat {
         let distX = point.x - self.x
         let distY = point.y - self.y
         let distance = sqrt(distX.ip_squared + distY.ip_squared)
         return distance
     }
 }
 
-public func +=(inout left: CGPoint, right: CGPoint) {
+public func +=(left: inout CGPoint, right: CGPoint) {
```
Comment (paulrolfe) [inline SwiftWisdom/Core/Geometry/Geometry.swift]: operator spacing

File: SwiftWisdom/Core/Geometry/Geometry.swift [hunk 5]
```diff
@@ -75,78 +75,63 @@ public func *(lhs: CGSize, rhs: CGFloat) -> CGSize {
...
 }
 
-public func random(range: Range<Int>) -> Int {
-    guard let first = range.first, let last = range.last else { return 0 }
-    assert(range.first >= 0)
+public func random(_ range: CountableClosedRange<Int>) -> Int {
+    let first = range.lowerBound
+    let last = range.upperBound
+
+    assert(first >= 0)
```
Comment (paulrolfe) [inline SwiftWisdom/Core/Geometry/Geometry.swift]: Why do we stop the range from being negative? That seems unexpected to me.

**Comment by brightredchilli** [comment]
actually, 4 unreviewed commits, the last 2 to update pods and add attributions are also technically unreviewed.

**Comment by Benuuu** [comment]
Looks like we didn't get dropping CompositeDisposable's key into this PR.  I will add it

File: SwiftWisdom/Core/StandardLibrary/NSAttributedString/NSMutableAttributedString+Format.swift [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            guard insertRange.location != NSNotFound else { return }
+            replaceCharacters(in: insertRange, with: arg)
+            rangeLimit = NSMakeRange(0, insertRange.location)
+            insertRange = str.range(of: "%@", options: .backwards, range: rangeLimit)
+        }
+    }
+}
+
+// TODO: move to String extensions before merging nyc/standard-library
+fileprivate extension String {
```
Comment (paulrolfe) [inline SwiftWisdom/Core/StandardLibrary/NSAttributedString/NSMutableAttributedString+Format.swift]: Is this TODO accurate? And any reason to keep this fileprivate? Seems like a useful extension to me.

**Review by Benuuu — APPROVED** [review]

File: SwiftWisdom/Core/StandardLibrary/Numbers/UnsignedInteger+Extensions.swift [hunk 7]
```diff
@@ -60,19 +60,25 @@ extension UnsignedIntegerType {
...
     public static var ip_maxValue: Self {
         return ip_bitStackOfLength(ip_maximumNumberOfBits)
     }
-    
-    public static var ip_maximumNumberOfBits: Self {
-        let size = UIntMax(sizeof(self))
-        return Self(size) * 8
+
+    // TODO: review!
+    public static var ip_maximumNumberOfBits: Int {
```
Comment (paulrolfe) [inline SwiftWisdom/Core/StandardLibrary/Numbers/UnsignedInteger+Extensions.swift:75]: This looks right, I don't think we need the TODO

File: SwiftWisdom/Core/StandardLibrary/Set/Set+Utilities.swift [hunk 8]
```diff
@@ -1,30 +1,31 @@
 
-public func +=<T>(inout lhs: Set<T>, rhs: Set<T>) {
+public func +=<T>(lhs: inout Set<T>, rhs: Set<T>) {
```
Comment (paulrolfe) [inline SwiftWisdom/Core/StandardLibrary/Set/Set+Utilities.swift]: add a space

File: SwiftWisdom/Core/StandardLibrary/Set/Set+Utilities.swift [hunk 9]
```diff
@@ -1,30 +1,31 @@
...
     lhs = lhs.union(rhs)
 }
 
-public func -=<T>(inout lhs: Set<T>, rhs: Set<T>) {
-    lhs = lhs.subtract(rhs)
+public func -=<T>(lhs: inout Set<T>, rhs: Set<T>) {
```
Comment (paulrolfe) [inline SwiftWisdom/Core/StandardLibrary/Set/Set+Utilities.swift]: add a space

File: SwiftWisdom/Core/StandardLibrary/String/String+Indexing.swift [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        return String(substringCharacters)
+    }
+
+    public subscript(range: CountableClosedRange<Int>) -> String {
+        let chars = Array(characters)
+        let substringCharacters = chars[range]
+        return String(substringCharacters)
+    }
+
+    public mutating func dropFirst() {
```
Comment (paulrolfe) [inline SwiftWisdom/Core/StandardLibrary/String/String+Indexing.swift]: Don't we need the `ip_` prefix on extensions?

File: SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+    func testContainsAll() {
+        let testArray = [1, 5, 9, 14, 12]
+        XCTAssert(testArray.ip_contains(all: [1, 12]))
+        XCTAssertFalse(testArray.ip_contains(all: [14, 16]))
+        XCTAssertFalse(testArray.ip_contains(all: [1, 5, 9, 10]))
+    }
+
+
+
```
Comment (paulrolfe) [inline SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift]: lots of extra white space at the bottom here

**Review by paulrolfe — APPROVED** [review]
Just a few nit picky things in my comments. I think this thing is pretty much good to go though. 👍

File: SwiftWisdom/Core/Async/Qu/Qu.swift (hunk 2 above)
Comment (brightredchilli) [reply SwiftWisdom/Core/Async/Qu/Qu.swift]: fixed

File: SwiftWisdom/Core/DirectoryManager/DirectoryManager.swift (hunk 3 above)
Comment (brightredchilli) [reply SwiftWisdom/Core/DirectoryManager/DirectoryManager.swift]: fixed and removed todo

File: SwiftWisdom/Core/Geometry/Geometry.swift (hunk 4 above)
Comment (brightredchilli) [reply SwiftWisdom/Core/Geometry/Geometry.swift]: fixed

File: SwiftWisdom/Core/Geometry/Geometry.swift (hunk 5 above)
Comment (brightredchilli) [reply SwiftWisdom/Core/Geometry/Geometry.swift]: i just realized that this code was being duplicated. We have implemented random(inRange:) somewhere else. ghastly. That other one doesnt have the assertion and is also tested. removing.

File: SwiftWisdom/Core/StandardLibrary/NSAttributedString/NSMutableAttributedString+Format.swift (hunk 6 above)
Comment (brightredchilli) [reply SwiftWisdom/Core/StandardLibrary/NSAttributedString/NSMutableAttributedString+Format.swift]: not accurate - moved this, and renamed it.

File: SwiftWisdom/Core/StandardLibrary/Numbers/UnsignedInteger+Extensions.swift (hunk 7 above)
Comment (brightredchilli) [reply SwiftWisdom/Core/StandardLibrary/Numbers/UnsignedInteger+Extensions.swift:75]: great. someone needs to document all of this stuff at some point haha

File: SwiftWisdom/Core/Geometry/Geometry.swift (hunk 4 above)
Comment (paulrolfe) [reply SwiftWisdom/Core/Geometry/Geometry.swift]: Is it?

**Pull request merged**